wagtail/models/__init__.py

```diff
@@ -2238,15 +2235,9 @@ class Revision(models.Model):
     page_revisions = PageRevisionsManager()
     submitted_revisions = SubmittedRevisionsManager()
 
-    @cached_property
-    def content_object(self):
-        return self.base_content_type.get_object_for_this_type(pk=self.object_id)
-
-    @cached_property
-    def specific_content_object(self):
-        if isinstance(self.content_object, Page):
-            return self.content_object.specific
-        return self.content_type.get_object_for_this_type(pk=self.object_id)
+    content_object = GenericForeignKey(
+        "content_type", "object_id", for_concrete_model=False
+    )
```

Comment:
This means that the `content_object` is always an instance of the specific page type. I don't think there's a downside to this as all tests seem to pass (after some minor adjustments in this PR).